client_cli/src/main.rs

```diff
@@ -111,27 +111,55 @@ pub enum Subcommand {
     Json(json::Args),
 }
 
+/// Context inside which command is executed
+pub trait RunContext {
+    /// Get access to configuration
+    fn configuration(&self) -> &ClientConfiguration;
+    /// Serialize and print data
+    ///
+    /// # Errors
+    /// - if serialization fails
+    /// - if printing fails
+    fn print_data(&mut self, data: &dyn Serialize) -> Result<()>;
+}
+
+struct PrintJsonContext<W> {
+    write: W,
+    config: ClientConfiguration,
+}
+
+impl<W: std::io::Write> RunContext for PrintJsonContext<W> {
+    fn configuration(&self) -> &ClientConfiguration {
+        &self.config
+    }
+
+    fn print_data(&mut self, data: &dyn Serialize) -> Result<()> {
```

Comment:
Why it should be `dyn Serialize` and not `impl Serialize`?

Reply:
It can be `impl Serialize` but in this case code size would be bigger because compiled code would have monomorphized version of function.

If in the future we would introduce another `context` like `PrintRONContext`. 
Size of code will be proportional to `n_contexts * m_serialized_objects`.

Atm difference is not that big through `9.2Mb` for generics vs `9.1Mb` for dyn object (`9.2Mb` vs `9.8Mb` if we would add `RON` serialization support).

Reply:
I think that for the client the compiled binary size is more important than a tiny performance benefit we would get with `impl`. So yeah, I am fine with `dyn`.